packages/routing-utils/src/superstatic.ts

```diff
@@ -211,91 +211,111 @@ export function collectHasSegments(has?: HasField) {
+    pathToRegexp(hash || '', hashKeys);
+    pathToRegexp(hostname || '', hostnameKeys);
+  } catch (_) {
+    // this is not fatal so don't error when failing to parse the
+    // params from the destination
+  }
 
-      destParams = new Set(
-        [...pathnameKeys, ...hashKeys]
-          .map(key => key.name)
-          .filter(val => typeof val === 'string') as string[]
-      );
+  destParams = new Set(
+    [...pathnameKeys, ...hashKeys, ...hostnameKeys]
+      .map(key => key.name)
+      .filter(val => typeof val === 'string') as string[]
+  );
 
-      pathname = safelyCompile(pathname, indexes, true);
-      hash = hash ? safelyCompile(hash, indexes, true) : null;
+  pathname = safelyCompile(pathname, indexes, true);
+  hash = hash ? safelyCompile(hash, indexes, true) : null;
+  hostname = hostname ? safelyCompile(hostname, indexes, true) : null;
 
-      for (const [key, strOrArray] of Object.entries(query)) {
-        if (Array.isArray(strOrArray)) {
-          query[key] = strOrArray.map(str => safelyCompile(str, indexes, true));
-        } else {
-          query[key] = safelyCompile(strOrArray, indexes, true);
-        }
-      }
+  for (const [key, strOrArray] of Object.entries(query)) {
+    if (Array.isArray(strOrArray)) {
+      query[key] = strOrArray.map(str =>
+        safelyCompile(unescapeSegments(str), indexes, true)
+      );
+    } else {
+      query[key] = safelyCompile(unescapeSegments(strOrArray), indexes, true);
     }
+  }
 
-    // We only add path segments to redirect queries if manually
-    // specified and only automatically add them for rewrites if one
-    // or more params aren't already used in the destination's path
-    const paramKeys = Object.keys(indexes);
-
-    if (!isRedirect && !paramKeys.some(param => destParams.has(param))) {
-      for (const param of paramKeys) {
-        if (!(param in query) && param !== UN_NAMED_SEGMENT) {
-          query[param] = indexes[param];
-        }
+  // We only add path segments to redirect queries if manually
+  // specified and only automatically add them for rewrites if one
+  // or more params aren't already used in the destination's path
+  const paramKeys = Object.keys(indexes);
+  const needsQueryUpdating =
+    !isRedirect && !paramKeys.some(param => destParams.has(param));
+
+  if (needsQueryUpdating) {
+    for (const param of paramKeys) {
+      if (!(param in query) && param !== UN_NAMED_SEGMENT) {
+        query[param] = indexes[param];
       }
     }
-
-    destination = formatUrl({
-      ...rest,
-      pathname,
-      query,
-      hash,
-    });
-
-    // url.format() escapes the dollar sign but it must be preserved for now-proxy
-    destination = destination.replace(/%24/g, '$');
   }
 
-  return destination;
+  destination = formatUrl({
+    ...rest,
+    hostname,
+    pathname,
+    query,
+    hash,
+  });
+
+  // url.format() escapes the dollar sign but it must be preserved for now-proxy
+  return destination.replace(/%24/g, '$');
```

Comment:
Does this replace need to happen at the bottom? I want to avoid replacing user input

Reply:
This is where this was happening previously, I don't think we can do this any earlier since it needs come after the `url.format`